### PR TITLE
Add support and test coverage for Swift 5

### PR DESCRIPTION
**Pull request by lacasseio**
### Context
Fixes https://github.com/gradle/gradle-native/issues/1002

### Gradle Core Team Checklist
- [ ] Verify design and implementation 
- [ ] Verify test coverage and [CI build status](https://builds.gradle.org/project.html?projectId=Gradle_Check&tab=projectOverview&branch_Gradle_Check=lacasseio%2Fswift-5)
- [ ] Verify documentation
- [ ] Recognize contributor in release notes

### CODE REVIEW

File: subprojects/platform-native/src/main/java/org/gradle/nativeplatform/toolchain/internal/metadata/CompilerMetaDataProvider.java [hunk 1]
```diff
@@ -20,11 +20,14 @@
 
 import java.io.File;
 import java.util.List;
+import java.util.Map;
 
 public interface CompilerMetaDataProvider<T extends CompilerMetadata> {
 
     SearchResult<T> getCompilerMetaData(File binary, List<String> additionalArgs, List<File> searchPath);
 
+    SearchResult<T> getCompilerMetaData(File binary, List<String> additionalArgs, List<File> searchPath, Map<String, String> environmentVariables);
```
Comment (lacasseio) [inline subprojects/platform-native/src/main/java/org/gradle/nativeplatform/toolchain/internal/metadata/CompilerMetaDataProvider.java]: ❓ This is by far the ugliest part of this PR. Any idea how to improve this? To switch between Xcode installation, the process needs to have `DEVELOPER_DIR` set to the installation, `.app`, directory. It's the official method, `xcrun` trick requires root access. We don't have any modelling within Gradle for Xcode to specify the `installDir`.

It's internal API, however I almost feel like we should be using a `Action<ExecSpec>` instead but I don't know how we could cache the call (inside `CachingCompilerMetaDataProvider`). Also, we don't allow all `ExecSpec` method to be configured. We could introduce a `CompilerExecSpec` that exposes a subset of the `ExecSpec` methods that allow us to extract the values and cache them. Something like:
```
interface CompilerExecSpec {
  CompilerExecSpec environment(String key, String value);
  CompilerExecSpec executable(Object executable);
  CompilerExecSpec args(String... args);
}
```
The prototype would become:
```
SearchResult<T> getCompilerMetaData(List<File> searchPath, Action<? super CompilerExecSpec> configureAction);
```

File: subprojects/language-native/src/main/java/org/gradle/language/nativeplatform/internal/toolchains/DefaultToolChainSelector.java [hunk 2]
```diff
@@ -109,6 +109,8 @@ static SwiftVersion toSwiftVersion(VersionNumber swiftCompilerVersion) {
             return SwiftVersion.SWIFT3;
```
Comment (big-guy) [inline subprojects/language-native/src/main/java/org/gradle/language/nativeplatform/internal/toolchains/DefaultToolChainSelector.java]: Just to remove one bit of maintenance...

```
    static SwiftVersion toSwiftVersion(VersionNumber swiftCompilerVersion) {
        for (SwiftVersion version : SwiftVersion.values()) {
            if (version.getVersion() == swiftCompilerVersion.getMajor()) {
                return version;
            }
        }
        throw new IllegalArgumentException(String.format("Swift language version is unknown for the specified Swift compiler version (%s)", swiftCompilerVersion.toString()));
    }
```

File: subprojects/platform-native/src/main/java/org/gradle/nativeplatform/toolchain/internal/swift/SwiftCompiler.java [hunk 3]
```diff
@@ -71,7 +72,7 @@ protected void addOptionsFileArgs(List<String> args, File tempDir) {
 
     @Override
     public WorkResult execute(SwiftCompileSpec spec) {
-        if (swiftCompilerVersion.getMajor() < spec.getSourceCompatibility().getVersion()) {
+        if (swiftCompilerVersion.getMajor() < spec.getSourceCompatibility().getVersion() || (swiftCompilerVersion.getMajor() >= 5 && spec.getSourceCompatibility().equals(SwiftVersion.SWIFT3))) {
```
Comment (big-guy) [inline subprojects/platform-native/src/main/java/org/gradle/nativeplatform/toolchain/internal/swift/SwiftCompiler.java]: WDYT about having a `SwiftVersion.isSupported(SwiftVersion)` where Swift 3 can support Swift 3, Swift 4 can support Swift 3&4 and Swift 5 can support Swift 4&5.

Comment (lacasseio) [reply subprojects/platform-native/src/main/java/org/gradle/nativeplatform/toolchain/internal/swift/SwiftCompiler.java]: It's a great idea, but I'm unsure we want the method to appear in the public API, anyway not yet. Alternatively, we could convert the `SwiftVersion` into an interface so we can have an internal API.

File: subprojects/language-native/src/main/java/org/gradle/language/nativeplatform/internal/toolchains/DefaultToolChainSelector.java (hunk 2 above)
Comment (lacasseio) [reply subprojects/language-native/src/main/java/org/gradle/language/nativeplatform/internal/toolchains/DefaultToolChainSelector.java]: Done.

**Comment by lacasseio** [comment]
All test pass on Xcode 10.2.

**Comment by big-guy** [comment]
@lacasseio could you look at the failures: https://builds.gradle.org/viewLog.html?buildId=21333582&tab=buildResultsDiv&buildTypeId=Gradle_Check_Stage_QuickFeedbackLinuxOnly_Trigger&branch_Gradle_Check=lacasseio%2Fswift-5

File: subprojects/platform-native/src/testFixtures/groovy/org/gradle/nativeplatform/fixtures/binaryinfo/OtoolBinaryInfo.groovy [hunk 4]
```diff
@@ -21,13 +21,15 @@ import org.gradle.nativeplatform.platform.internal.Architectures
 
 class OtoolBinaryInfo implements BinaryInfo {
     def binaryFile
+    private final List<String> environments
 
-    OtoolBinaryInfo(File binaryFile) {
+    OtoolBinaryInfo(File binaryFile, List<String> environments) {
```
Comment (big-guy) [inline subprojects/platform-native/src/testFixtures/groovy/org/gradle/nativeplatform/fixtures/binaryinfo/OtoolBinaryInfo.groovy]: Is this actually needed for running dwarfdump now?  Is this related to DEVELOPER_DIR?

File: subprojects/platform-native/src/main/java/org/gradle/nativeplatform/toolchain/internal/metadata/CompilerMetaDataProvider.java [hunk 5]
```diff
@@ -16,15 +16,22 @@
...
+import org.gradle.api.Action;
 import org.gradle.platform.base.internal.toolchain.SearchResult;
 
 import java.io.File;
 import java.util.List;
 
 public interface CompilerMetaDataProvider<T extends CompilerMetadata> {
 
-    SearchResult<T> getCompilerMetaData(File binary, List<String> additionalArgs, List<File> searchPath);
+    SearchResult<T> getCompilerMetaData(List<File> searchPath, Action<? super CompilerExecSpec> configureAction);
```
Comment (big-guy) [inline subprojects/platform-native/src/main/java/org/gradle/nativeplatform/toolchain/internal/metadata/CompilerMetaDataProvider.java]: I'm not quite sure I understand why this is necessary.  Does DEVELOPER_DIR need to be set in order to run the compiler to get information about it?

File: subprojects/platform-native/src/main/java/org/gradle/nativeplatform/toolchain/internal/xcode/AbstractLocator.java [hunk 6]
```diff
@@ -41,6 +41,11 @@ public File find() {
                 execAction.executable("xcrun");
                 execAction.workingDir(System.getProperty("user.dir"));
                 execAction.args(getXcrunFlags());
+
+                String developerDir = System.getenv("DEVELOPER_DIR");
+                if (developerDir != null) {
+                    execAction.environment("DEVELOPER_DIR", developerDir);
+                }
```
Comment (big-guy) [inline subprojects/platform-native/src/main/java/org/gradle/nativeplatform/toolchain/internal/xcode/AbstractLocator.java]: I think we inherit the environment from the parent process, is this necessary?

File: subprojects/platform-native/src/main/java/org/gradle/nativeplatform/toolchain/internal/swift/SwiftPlatformToolProvider.java [hunk 7]
```diff
@@ -129,6 +129,11 @@ private CommandLineToolInvocationWorker commandLineTool(ToolType key, String exe
     private CommandLineToolContext context(CommandLineToolConfigurationInternal toolConfiguration) {
         MutableCommandLineToolContext baseInvocation = new DefaultMutableCommandLineToolContext();
         baseInvocation.setArgAction(toolConfiguration.getArgAction());
+
+        String developerDir = System.getenv("DEVELOPER_DIR");
```
Comment (big-guy) [inline subprojects/platform-native/src/main/java/org/gradle/nativeplatform/toolchain/internal/swift/SwiftPlatformToolProvider.java]: Can this be derived from the tool installation directory?

e.g., instead of having to make sure the DEVELOPER_DIR environment variable is set properly, could we derive what DEVELOPER_DIR should be or just add something to SwiftcToolChain to set it when we configure the swift tool chain from AvailableToolChains?

It seems like that might make some of this simpler since we don't have to pass around things through environment variables.

File: subprojects/platform-native/src/main/java/org/gradle/nativeplatform/toolchain/internal/metadata/CompilerMetaDataProvider.java (hunk 5 above)
Comment (lacasseio) [reply subprojects/platform-native/src/main/java/org/gradle/nativeplatform/toolchain/internal/metadata/CompilerMetaDataProvider.java]: The `man xcode-select` have a section that mentions every tool in the following list depends on the `DEVELOPER_DIR` (or `xcode-select`) to switch version:
```
FILES
       /usr/bin/xcrun
          Used to find or run arbitrary commands from the active developer directory. See xcrun(1) for more information.

       /usr/bin/actool  /usr/bin/agvtool  /usr/bin/desdp  /usr/bin/genstrings  /usr/bin/ibtool /usr/bin/ictool /usr/bin/instruments /usr/bin/iprofiler /usr/bin/opendiff /usr/bin/sdef /usr/bin/sdp /usr/bin/stapler
       /usr/bin/xcodebuild /usr/bin/xed
          Runs the matching Xcode tool from with the active developer directory.

       /usr/bin/BuildStrings /usr/bin/CpMac /usr/bin/DeRez /usr/bin/GetFileInfo /usr/bin/MergePef /usr/bin/MvMac /usr/bin/ResMerger /usr/bin/RezDet /usr/bin/RezWack /usr/bin/Rez  /usr/bin/SetFile  /usr/bin/Split-
       Forks   /usr/bin/UnRezWack  /usr/bin/ar  /usr/bin/as  /usr/bin/asa  /usr/bin/bison  /usr/bin/c89  /usr/bin/c99  /usr/bin/clang++  /usr/bin/clang  /usr/bin/cmpdylib  /usr/bin/codesign_allocate  /usr/bin/cpp
       /usr/bin/ctags  /usr/bin/ctf_insert  /usr/bin/dsymutil  /usr/bin/dwarfdump  /usr/bin/flex++   /usr/bin/flex   /usr/bin/g++   /usr/bin/gatherheaderdoc   /usr/bin/gcc   /usr/bin/gcov   /usr/bin/git-cvsserver
       /usr/bin/git-receive-pack  /usr/bin/git-shell  /usr/bin/git-upload-archive  /usr/bin/git-upload-pack  /usr/bin/git /usr/bin/gm4 /usr/bin/gnumake /usr/bin/gperf /usr/bin/hdxml2manxml /usr/bin/headerdoc2html
       /usr/bin/indent /usr/bin/install_name_tool /usr/bin/ld /usr/bin/lex  /usr/bin/libtool  /usr/bin/lipo  /usr/bin/lldb  /usr/bin/lorder  /usr/bin/m4  /usr/bin/make  /usr/bin/mig  /usr/bin/mkdep  /usr/bin/nasm
       /usr/bin/ndisasm   /usr/bin/nm   /usr/bin/nmedit   /usr/bin/objdump   /usr/bin/otool   /usr/bin/pagestuff  /usr/bin/ranlib  /usr/bin/rebase  /usr/bin/redo_prebinding  /usr/bin/resolveLinks  /usr/bin/rpcgen
       /usr/bin/segedit /usr/bin/size /usr/bin/strings /usr/bin/strip /usr/bin/svn /usr/bin/svnadmin /usr/bin/svndumpfilter /usr/bin/svnlook /usr/bin/svnserve /usr/bin/svnsync  /usr/bin/svnversion  /usr/bin/swift
       /usr/bin/swiftc /usr/bin/unifdef /usr/bin/unifdefall /usr/bin/xml2man /usr/bin/yacc
          Runs the matching BSD tool from with the active developer directory.
```

The tools are in a single location but redirect to the right Xcode installation. We can't use `xcode-select` as concurrent execution would cause the flakiness. To query the right metadata, the compiler needs to be invoked using the right `DEVELOPER_DIR` variable. The compiler (the official way from my understanding) is always located in the same directory so we can't infer the right value from the compiler path. The changes have effects only for our testing infrastructure at the moment. It also has the side-effect of switching compiler version if someone starts Gradle with that environment variable set. I kept it internal and dependent on the environment variable for now as we supporting a public API needs more work. In the future, we could see the tool chain discovery able to manager Xcode version properly. Right now, if you only set the path to the compiler, you may still get incorrect result as `xcrun` and co. will assume the default Xcode (from `/Applications/Xcode.app`).

I assumed the worker API wouldn't carry the `DEVELOPPER_DIR` if it fork the work to another process so I pass the value were it's required.

File: subprojects/platform-native/src/main/java/org/gradle/nativeplatform/toolchain/internal/swift/SwiftPlatformToolProvider.java (hunk 7 above)
Comment (lacasseio) [reply subprojects/platform-native/src/main/java/org/gradle/nativeplatform/toolchain/internal/swift/SwiftPlatformToolProvider.java]: I don't know enough about the worker API and how they manage the environment variable. Do you know if the variables are passed to the workers?

File: subprojects/platform-native/src/main/java/org/gradle/nativeplatform/toolchain/internal/xcode/AbstractLocator.java (hunk 6 above)
Comment (lacasseio) [reply subprojects/platform-native/src/main/java/org/gradle/nativeplatform/toolchain/internal/xcode/AbstractLocator.java]: We may, I wanted to be explicit so we know where to wire things properly when we support the feature on the tool chain. Do you think we should remove as much as we can?

File: subprojects/platform-native/src/testFixtures/groovy/org/gradle/nativeplatform/fixtures/binaryinfo/OtoolBinaryInfo.groovy (hunk 4 above)
Comment (lacasseio) [reply subprojects/platform-native/src/testFixtures/groovy/org/gradle/nativeplatform/fixtures/binaryinfo/OtoolBinaryInfo.groovy]: I prefer using the tools from the Xcode we used to build. This is required to use the right tool version.
